zipline/utils/exchange_calendar.py

```diff
@@ -0,0 +1,386 @@
+#
+# Copyright 2016 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import (
+    ABCMeta,
+    abstractproperty,
+    abstractmethod,
+)
+from six import with_metaclass
+
+import pandas as pd
+from pandas import (
+    DataFrame,
+    date_range,
+    DateOffset,
+    DatetimeIndex,
+    Timedelta,
+)
+from pandas.tseries.offsets import CustomBusinessDay
+
+start_default = pd.Timestamp('1990-01-01', tz='UTC')
+end_base = pd.Timestamp('today', tz='UTC')
+# Give an aggressive buffer for logic that needs to use the next trading
+# day or minute.
+end_default = end_base + pd.Timedelta(days=365)
+
+
+def delta_from_time(t):
+    """
+    Convert a datetime.time into a timedelta.
+    """
+    return Timedelta(
+        hours=t.hour,
+        minutes=t.minute,
+        seconds=t.second,
+    )
+
+
+def days_at_time(days, t, tz, day_offset=0):
+    """
+    Shift an index of days to time t, interpreted in tz.
+
+    Overwrites any existing tz info on the input.
+
+    Parameters
+    ----------
+    days : DatetimeIndex
+        The "base" time which we want to change.
+    t : datetime.time
+        The time we want to offset @days by
+    tz : pytz.timezone
+        The timezone which these times represent
+    day_offset : int
+        The number of days we want to offset @days by
+    """
+    days = DatetimeIndex(days).tz_localize(None).tz_localize(tz)
```

Comment:
This is so we can easily flip from midnight UTC to midnight EST. We index our schedule in UTC, but in order to offset correctly given that `NYSE_OPEN` and `NYSE_CLOSE` are timezone naive, we must my convert our dates into EST: UTC -> Naive -> EST.